core/src/test/java/smile/projection/ICATest.java

```diff
@@ -0,0 +1,167 @@
+        try {
+            AttributeDataset mixed = parser
+                    .parse(smile.data.parser.IOUtils.getTestDataFile("projection/mixed-signal/test_ica_X.csv"));
+            double[][] x = mixed.toArray(new double[mixed.size()][]);
+
+            parser.setResponseIndex(new NominalAttribute("class"), 1);
+            AttributeDataset signal1 = parser
+                    .parse(smile.data.parser.IOUtils.getTestDataFile("projection/mixed-signal/test_ica_s1.csv"));
+            double[][] s1 = signal1.toArray(new double[signal1.size()][]);// sinusoidal
+
+            AttributeDataset signal2 = parser
+                    .parse(smile.data.parser.IOUtils.getTestDataFile("projection/mixed-signal/test_ica_s2.csv"));
+            double[][] s2 = signal2.toArray(new double[signal2.size()][]);// square
+
+            AttributeDataset signal3 = parser
+                    .parse(smile.data.parser.IOUtils.getTestDataFile("projection/mixed-signal/test_ica_s3.csv"));
+            double[][] s3 = signal3.toArray(new double[signal3.size()][]);// sawtooth
+
+            ICA fastICA = new ICA(x, 3);
+
+            double[][] p = fastICA.project(x);
+            int size = mixed.size();
+
+            double[] p1 = new double[size];// sawtooth
+            for (int i = 0; i < size; i++) {
+                p1[i] = p[i][0];
+            }
+
+            double[] p2 = new double[size];// square
+            for (int i = 0; i < size; i++) {
+                p2[i] = p[i][1];
+            }
+
+            double[] p3 = new double[size];// sinusoidal
+            for (int i = 0; i < size; i++) {
+                p3[i] = p[i][2];
+            }
+
+            double[] o1 = new double[size];
+            for (int i = 0; i < size; i++) {
+                o1[i] = s1[i][0];
+            }
+
+            double[] o2 = new double[size];
+            for (int i = 0; i < size; i++) {
+                o2[i] = s2[i][0];
+            }
+
+            double[] o3 = new double[size];
+            for (int i = 0; i < size; i++) {
+                o3[i] = s3[i][0];
+            }
+
+            double mse1 = 0;
+            for (int i = 0; i < size; i++) {
+                double diff = p1[i] - o3[i];
+                mse1 += diff * diff;
+            }
+            double rmse1 = Math.sqrt(mse1 / size);
+            System.out.println("rmse for signal1 is " + rmse1);
+            assertTrue(rmse1 < 1);
+
+            double mse2 = 0;
+            for (int i = 0; i < size; i++) {
+                double diff = p2[i] - o2[i];
+                mse2 += diff * diff;
+            }
+            double rmse2 = Math.sqrt(mse2 / size);
+            System.out.println("rmse for signal2 is " + rmse2);
+            assertTrue(rmse2 < 1);
+
+            double mse3 = 0;
+            for (int i = 0; i < size; i++) {
+                double diff = p3[i] - o1[i];
+                mse3 += diff * diff;
+            }
+            double rmse3 = Math.sqrt(mse3 / size);
```

Comment:
Not sure RMSE is the right metric to measure ICA.

Reply:
@haifengl rmse here is used here as a quick and rough comparison taken point-by-point. Yes, for ICA, I think the most intuitive way is to check the visualization of unmixed-signals which also shows the power of ICA to distinguish mixed signals.